controllers/crds/cninode_controller.go

```diff
@@ -0,0 +1,194 @@
+			return ctrl.Result{}, fmt.Errorf("failed to patch CNINode %v", err)
+		}
+
+		// Re-fetch the CNINode after updating the Tag so we have the latest state of the CRD before updating it again in later operations
+		if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Add the finalizer if it does not exist, so we can define cleanup routine which should be run before CNINode can be deleted
+	if !controllerutil.ContainsFinalizer(cniNode, config.NodeTerminationFinalizer) {
+		r.Log.Info("adding finalizer for CNINode", "cniNode", cniNode.Name)
+		if ok := controllerutil.AddFinalizer(cniNode, config.NodeTerminationFinalizer); !ok {
+			r.Log.Info("failed to add finalizer for CNINode, requeuing the request", "cniNode", cniNode.Name)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		if err := r.Client.Update(ctx, cniNode); err != nil {
+			// Requeue request so update can be retried
+			return ctrl.Result{}, err
+		}
+	}
+
+	if cniNode.GetDeletionTimestamp() != nil {
+		// CNINode is marked for deletion
+		if controllerutil.ContainsFinalizer(cniNode, config.NodeTerminationFinalizer) {
+			// check if node object exists
+			node := &v1.Node{}
+			if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+				if errors.IsNotFound(err) {
+					//  node is also deleted, proceed with running the cleanup routine and remove the finalizer
+					r.Log.Info("running the finalizer routine on cniNode", "cniNode", cniNode.Name)
+					cleaner := &cleanup.NodeTerminationCleaner{
+						NodeName: cniNode.Name,
+					}
+					cleaner.ENICleaner = &cleanup.ENICleaner{
+						EC2Wrapper: r.EC2Wrapper,
+						Manager:    cleaner,
+						VPCID:      r.VPCID,
+						Log:        ctrl.Log.WithName("eniCleaner").WithName("node"),
+					}
+					// Return err if failed to delete leaked ENIs on node so it can be retried
+					if err := cleaner.DeleteLeakedResources(); err != nil {
+						return ctrl.Result{}, err
+					}
+					// Remove finalizer and update CNINode to continue deletion
+					if ok := controllerutil.RemoveFinalizer(cniNode, config.NodeTerminationFinalizer); !ok {
+						r.Log.Info("failed to remove finalizer for CNINode, requeuing the request", "cniNode", cniNode.Name)
+						return ctrl.Result{Requeue: true}, nil
+					}
+					if err := r.Client.Update(ctx, cniNode); err != nil {
+						// Return err so update can be retried
+						return ctrl.Result{}, err
+					}
+				} else {
+					r.Log.Info("failed to get the node object in CNINode reconciliation, will retry")
+					// Requeue request so it can be retried
+					return ctrl.Result{}, err
+				}
+			} else {
+				// node exists, do not run the cleanup routine(periodic cleanup routine will anyway delete leaked ENIs), remove the finalizer
+				// and let object be deleted and re-create similar object
+
+				// Create a copy without deletion timestamp for creation
+				newCNINode := &v1alpha1.CNINode{
```

Comment:
Can you use `newCNINode := cniNode.DeepCopy()`?

Reply:
Yeah, this is not directly settable by clients per documentation: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata